include/libcyphal/presentation/shared_object.hpp

```diff
@@ -25,9 +25,53 @@ namespace presentation
 namespace detail
 {
 
+/// @brief Defines double-linked list of unreferenced nodes.
+///
+struct UnRefNode
+{
+    void linkAsUnreferenced(UnRefNode& origin)
+    {
+        CETL_DEBUG_ASSERT((prev_node != nullptr) == (next_node != nullptr), "Should be both or none.");
+
+        // Already linked?
+        if ((nullptr == prev_node) && (nullptr == next_node))
+        {
+            // Link to the end of the list, so that the object is destroyed in the order of unreferencing.
+            //
+            next_node                   = &origin;
+            prev_node                   = origin.prev_node;
+            origin.prev_node->next_node = this;
+            origin.prev_node            = this;
+        }
+    }
+
+    void unlinkIfReferenced()
+    {
+        CETL_DEBUG_ASSERT((prev_node != nullptr) == (next_node != nullptr), "Should be both or none.");
+
+        // Already unlinked?
+        if ((nullptr != prev_node) && (nullptr != next_node))
+        {
+            prev_node->next_node = next_node;
+            next_node->prev_node = prev_node;
+
+            next_node = nullptr;
+            prev_node = nullptr;
+        }
+    }
+
+    // No Lint b/c this `UnRefNode` is a simple helper struct as base of the below `SharedObject`.
+    // It's under `detail` namespace and not supposed to be used directly by the users of the library.
+    // NOLINTBEGIN(misc-non-private-member-variables-in-classes)
+    UnRefNode* prev_node{nullptr};
```

Comment:
could we protect these by making them private then using `friend class SharedObject`?

Reply:
we probably can but I'm not sure we need - all derived classes (from `SharedObject`) are still "Impl" internal (and under `detail` namespace) entities. F.e. if you work with `Publisher<Message>` (which is not shared object!) then you don't see under the hood involved `PublisherImpl`). The same story for 2 other entities: subscriber and rpc client.